Mathlib/Data/List/Basic.lean

```diff
@@ -1812,6 +1812,40 @@ theorem takeWhile_takeWhile (p q : α → Bool) (l : List α) :
 theorem takeWhile_idem : takeWhile p (takeWhile p l) = takeWhile p l := by
   simp_rw [takeWhile_takeWhile, and_self_iff, Bool.decide_coe]
 
+variable (p) (l)
+
+lemma find?_eq_head?_dropWhile_not :
+    l.find? p = (l.dropWhile (fun x ↦ ! (p x))).head? := by
+  induction l
+  case nil => simp
+  case cons head tail hi =>
+    set ph := p head with phh
+    rcases ph with rfl | rfl
+    · have phh' : ¬(p head = true) := by simp [phh.symm]
+      rw [find?_cons_of_neg _ phh', dropWhile_cons_of_pos]
+      · exact hi
+      · simpa using phh
+    · rw [find?_cons_of_pos _ phh.symm, dropWhile_cons_of_neg]
+      · simp
+      · simpa using phh
+
+lemma find?_not_eq_head?_dropWhile :
+    l.find? (fun x ↦ ! (p x)) = (l.dropWhile p).head? := by
+  convert l.find?_eq_head?_dropWhile_not ?_
+  simp
+
+variable {p} {l}
+
+lemma find?_eq_head_dropWhile_not (h : l.dropWhile (fun x ↦ ! (p x)) ≠ []) :
```

Comment:
This is fine, but I'm curious if these lemmas would be more or less useful if we replaced this hypothesis with
`\exists x \mem l, p x`, and then transformed that into the proof that `.head` expects in the theorem statement.

Reply:
Since I'm only using one of these lemmas (this one), in one place in #18685 (and provided all four lemmas since they naturally seem to go together), I don't have a clear sense of exactly what form of statement would be most useful in general.

Reply:
Oh, then I think you should change it, as at your use [site](https://github.com/leanprover-community/mathlib4/pull/18685/files#diff-192f38c2de762046d78c17d8dd16c6c45474707c0dc654d23556ee2f611f6ba0R384) things would get easier, i.e. you wouldn't need to call `dropWhile_eq_nil_iff` to prepare the proof.

Reply:
OK, I've made the change (it doesn't make things much simpler at the use site since the original hypothesis is still needed in other places in that proof to pass to other lemmas).